### PR TITLE
feat(cmake): run busted tests with ctest

### PR DESCRIPTION
**Pull request by kylo252**
## Description

Make use of [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html) to streamline running busted tests\tasks

### Features

- add `busted_discover_tests` which enables running _each_ test in an isolated environment, with a nice short-hand
  ```console
  TEST_FILE=test/functional/api/autocmd_spec.lua TEST_FILTER='removes.an.autocommand' cmake --build build --target functionaltest
  # vs
  ctest --test-dir build -R 'removes.an.autocommand'
  ```
- add the ability to run tests by [labels](https://github.com/kylo252/neovim/blob/978649c15ea3269fa8d60bdfd69b31c77eac8df8/.github/workflows/ci.yml#L268-L280)

#### Additional improvements

- use the configuration file [.busted](https://github.com/Olivine-Labs/busted/blob/0fc1f4faedcfee287fb925fec78ef88488a2df3a/spec/.hidden/.busted) to set common flags needed to be passed to `busted` on every run
- create a single `test/preload.lua` that can detect which specific helper to load using based on args passed to `-Xhelper`

### How to use

- run all the tests
  ```sh
  ctest --test-dir build --progress
  # or
  cmake --build build --target test
  # or
  ninja -C build test
  ```
- list all available labels
  ```sh
  ctest --test-dir build --print-labels
  ```
- list all available tests
  ```sh
  ctest --test-dir build --list
  ```
- run tests by label `-L or --label-regex`, see all available [flags](https://cmake.org/cmake/help/latest/manual/ctest.1.html#options)
  ```sh
  ctest --test-dir build -L 'api_spec'
  ```

### CODE REVIEW

**Comment by justinmk** [comment]
> Using the configuration file [.busted](../.busted) from #17090 will simplify the process, by reducing the number of arguments needed to be passed to `busted` on every run

Developers don't run `busted` directly, so that is not relevant.

**Comment by kylo252** [comment]
> Developers don't (shouldn't, usually) run `busted` directly, so that is not relevant.

the main reason I care about this config file is test-suit capability, the way I see it:
- developers should be able to run incremental testing without too much fiddling -> "run all api tests"
- it would be nice if the CI can selectively turn off test-fixtures without hard-coding `pending` every where, e.g. https://github.com/neovim/neovim/pull/18515

**Comment by justinmk** [comment]
> developers should be able to run incremental testing without too much fiddling -> "run all api tests"

That is already possible via TEST_FILTER and friends.

> it would be nice if the CI can selectively turn off test-fixtures without hard-coding pending every where

Adding another layer isn't a desirable way to do that.

File: cmake/RunTests.cmake [hunk 1]
```diff
@@ -20,12 +20,16 @@ if(NVIM_PRG)
   set(ENV{NVIM_PRG} "${NVIM_PRG}")
 endif()
 
-if(DEFINED ENV{TEST_FILE})
+if(TEST_PATH)
+  set(TEST_PATH "${TEST_PATH}")
+elseif(DEFINED ENV{TEST_FILE})
```
Comment (kylo252) [inline cmake/RunTests.cmake:25]: > Adding another layer isn't a desirable way to do that.

@justinmk, until we agree on that, what do you think of re-write parts of `RunTest.cmake` to work with this then?

Comment (justinmk) [reply cmake/RunTests.cmake:25]: if the rewrite leads to something useful then we want that :) If it leads to churn without a purpose then I want to save you time by cautioning against that.

Comment (kylo252) [reply cmake/RunTests.cmake:25]: as this is still a very rough draft, and since you don't seem that convinced, I'll leave it open for a while.

**Comment by justinmk** [comment]
> The idea is to consolidate (or replace) [cmake/RunTests.cmake](https://github.com/neovim/neovim/cmake/RunTests.cmake) with a proper wrapper to streamline running tests.

What is a concrete example of how it would be better streamlined? Would GHA be able to parse the tests automatically (that would be a clear gain)? Does it help set the stage for parallel tests? Does it give us a _net reduction_ in code?

**Comment by kylo252** [comment]
> > The idea is to consolidate (or replace) [cmake/RunTests.cmake](https://github.com/neovim/neovim/cmake/RunTests.cmake) with a proper wrapper to streamline running tests.
> 
> What is a concrete example of how it would be better streamlined? 

there would be a single way to run and evaluate tests, see [PASS_REGULAR_EXPRESSION](https://cmake.org/cmake/help/v3.9/prop_test/PASS_REGULAR_EXPRESSION.html)

no more 
https://github.com/neovim/neovim/blob/e7b3fd8ad603ba064ce9457a8866430c436c3737/ci/common/suite.sh#L8-L18

or

https://github.com/neovim/neovim/blob/e7b3fd8ad603ba064ce9457a8866430c436c3737/ci/build.ps1#L106-L108

> Would GHA be able to parse the tests automatically (that would be a clear gain)? 

busted should be able to do this natively, I talked to @dundargoc about this, I'm not sure why it's not configured as such. so we need to configure either of them to use JUnit as output, and then hook them up to one of the many available parsers, e.g. https://github.com/mikepenz/action-junit-report or https://github.com/EnricoMi/publish-unit-test-result-action

It also has a `ctest_memcheck()` command, although not sure how useful it is, but apparently it can dump the results in an xml format, ref: https://cmake.org/cmake/help/v3.9/command/ctest_memcheck.html#command:ctest_memcheck

> Does it help set the stage for parallel tests? 

yup, this is one of the clear benefits of CTest, ref: https://cmake.org/cmake/help/v3.9/manual/ctest.1.html#options

> Does it give us a _net reduction_ in code? 

that's a tricky one to answer, it shouldn't be more than a single CMake module/file

> Is it better supported by test plugins?

not sure what you mean by plugins, but they have a big integration with something called [CDash](https://github.com/Kitware/CDash) if you find that useful, some examples I found:
- https://my.cdash.org/index.php?project=libevent
- https://my.cdash.org/index.php?project=libarchive
- https://my.cdash.org/index.php?project=jsoncpp

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    COMMAND ${CMAKE_COMMAND}
+    -DBUSTED_PRG=${BUSTED_PRG}
+    -DLUA_PRG=${LUA_PRG}
+    -DWORKING_DIR=${CMAKE_CURRENT_SOURCE_DIR}
+    -DBUSTED_OUTPUT_TYPE=${BUSTED_OUTPUT_TYPE}
+    -DTEST_DIR=${CMAKE_CURRENT_SOURCE_DIR}/test
+    -DBUILD_DIR=${CMAKE_BINARY_DIR}
+    -DTEST_TYPE=unit
+    -P ${PROJECT_SOURCE_DIR}/cmake/RunTests.cmake
+    DEPENDS ${UNITTEST_PREREQS}
```
Comment (andmis) [inline test/CMakeLists.txt:48]: Shouldn't you just depend on the `unittest-prereqs` target here? And in that case the three local variables `[UNITTEST|FUNCTIONALTEST|BENCHMARK]_PREREQS` are not actually needed, the lists of prereqs can be inlined in the `add_custom_target` calls.

File: test/preload.lua [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    test_type = 'functional'
+  elseif value:match('IS_UNIT_TEST') then
+    test_type = 'unit'
+  elseif value:match('IS_BENCHMARK_TEST') then
+    test_type = 'benchmark'
+  end
+end
+
+local luv = require('luv')
+-- NOTE: os_setenv() isn't thread-safe according to luv's docs
```
Comment (justinmk) [inline test/preload.lua:17]: we aren't using threads in the test runner

**Comment by kylo252** [comment]
### Status

- it seems that the current test failures are related to environment isolation,  probably `$XDG_STATE_HOME`, rather than CTest itself, see https://github.com/neovim/neovim/runs/6988757128?check_suite_focus=true#step:17:3982
- running individual tests (collected by `busted_discover_tests`) has a noticeable slowdown when running below `-j4` 
	- only enable this mode for a single job, since the CI doesn't have enough threads
	- make it clear for devs that they need to either pass `-j` or set `$CTEST_PARALLEL_LEVEL`
	- add test-suites which runs multiple tests serially but we run the suites themselves in parallel, the speed up is about 200-300%

File: .busted [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  unit = {
+    ROOT = { 'test/unit' },
+    Xhelper = { 'IS_UNIT_TEST' },
+  },
+  ---
+  benchmark = {
+    ROOT = { 'test/benchmark' },
+    Xhelper = { 'IS_BENCHMARK_TEST' },
+  },
+}
```
Comment (justinmk) [inline .busted:37]: Is this file required for this PR?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -260,6 +260,28 @@ jobs:
         name: Install nvim
         run: ./ci/run_tests.sh install_nvim
 
+      - if: matrix.flavor != 'tsan' && matrix.flavor != 'functionaltest-lua' && !cancelled()
+        name: Generate test-runners
+        run: |
+          cmake --build $GITHUB_WORKSPACE/build --target generate-test-runners
+
+
+      # macOS runners have 3 threads
```
Comment (justinmk) [inline .github/workflows/ci.yml:269]: Why are we fiddling with threads? What evidence is there that the test runner is CPU-bound (not IO-bound)? How much faster is the test run with parallel invocation?

File: .gitignore [hunk 6]
```diff
@@ -73,3 +73,6 @@ tags
 
 # vim patches
 /vim-*.patch
+
+# ctest artifcats
+Testing
```
Comment (justinmk) [inline .gitignore:78]: should this go in `build/` (maybe `build/test/` or something)?

File: test/functional/helpers.lua [hunk 7]
```diff
@@ -462,6 +462,10 @@ function module.new_argv(...)
         'NVIM_RPLUGIN_MANIFEST',
         'GCOV_ERROR_FILE',
         'XDG_DATA_DIRS',
+        'XDG_CONFIG_HOME',
+        'XDG_DATA_HOME',
+        'XDG_STATE_HOME',
+        'XDG_CACHE_HOME',
```
Comment (justinmk) [inline test/functional/helpers.lua:468]: if this is needed, should be a separate PR

**Review by justinmk — CHANGES_REQUESTED** [review]

File: .busted (hunk 4 above)
Comment (kylo252) [reply .busted:37]: It's a bit tricky to say one or another, so I'll try to summarize a list of pros and cons

### pros
- debugging: since you need very few, or no arguments at all, to run tests `busted path/to/test`
- ergonomics(1): no need create a [custom argparse](https://github.com//neovim/neovim/blob/b7084fef4c850d0352488b14dcff0f36a7e75e1c/cmake/RunTests.cmake#L35-L40) that determines what is safe to override (`--output`) and what isn't (`--lpath`)
- ergonomics(2): no need to [expose environment variables](https://github.com//neovim/neovim/blob/b7084fef4c850d0352488b14dcff0f36a7e75e1c/cmake/RunTests.cmake#L42-L52) either instead of the aforementioned argparse

### cons
- it may seem that running tests directly with busted works otob, you still need to set some environment variables (this used to be a lot more involved before I've delegated a lot of these responsibilities to [`preload.lua`](https://github.com//kylo252/neovim/blob/cbdd5a9662da6c061755682e58d18416e204c946/test/preload.lua#L136-L166))
- bloat(?): always a consideration when adding yet another file to the repo, although it can actually go under `./test`

File: .github/workflows/ci.yml (hunk 5 above)
Comment (kylo252) [reply .github/workflows/ci.yml:269]: on the CI, this is a bit harder to gauge, but here are the stats on my machine

<table>
<tr>
	<td>cmd
	<td>time
<tr>
	<td>TEST_FILE=test/functional/api cmake --build build --target functionaltest
	<td> 11s
<tr>
	<td>ctest --test-dir build -L 'spec' --repeat until-pass:3 -j8 --progress
	<td>8.75s
<tr>
	<td>ctest --test-dir build -L 'spec' --repeat until-pass:3 -j16 --progress
	<td>5.85s
</table>

File: .gitignore (hunk 6 above)
Comment (kylo252) [reply .gitignore:78]: I'm not sure I can control where this folder is created, but I can look it up to see if there's a way around it.

probably related: https://gitlab.kitware.com/cmake/cmake/-/issues/18156
